docs/synapse/model_format.rst

```diff
@@ -0,0 +1,234 @@
+
+  #. The name of the model to be updated.
+  #. The timestamp of the model revision, in integer form.  This is validated using the "%Y%m%d%H%M" format string.
+
+These decorated functions should implement any changes neccesary for the data model; and perform any neccesary data
+migrations.  For performance reasons, these migrations should be performed using the **storage layer** APIs.  The use
+of node (tufo) level APIs when doing data migrations may result in terrible performance, causing migrations to take
+a significant amount of time.
+
+Under the hood, when the model is loaded by the Cortex, the rev0 function and any @modelrev decorated functions are
+sorted by model version, and these functions are then executed when CoreModule initialization is completed. This has
+the following effects:
+
+    - On new Cortex's, the rev0 function will be loaded (adding the complete data model). Then subsequent functions
+      will be executed.  These should perform any model changes and data migrations as needed.
+    - On existing Cortexes which have already had the base model loaded; the rev0 function will be skipped and
+      subsequent functions executed until the last function has been executed.  These functions should perform a model
+      addition of the basemodel contents, and perform any required migration actions using the **storage layer** APIs.
+
+The following is a second revision of the earlier FooBarModule - it add a property for the foo:bar type, and performs
+a storage layer migration of nodes to set a default value for the new property. The model revision for the foobar model
+will be updated to 201707210101 in the cortex.  Note that during the model revision function, we exit the function
+early if the new property already exists.  This would cause the current model version to be set and any subsequent
+functions called.::
+
+    import synapse.lib.module as s_module
+
+    def getDataModel():
+        return FooBarModule.getBaseModels()[0][1]
+
+    class FooBarModule(s_module.CoreModule):
+
+        @staticmethod
+        def getBaseModels():
+            modl = {
+                'types': (
+                    ('foo:bar', {'subof': 'str', 'doc': 'A foo bar!'}),
+                ),
+                'forms': (
+                    ('foo:bar',
+                     {'ptype': 'str'},
+                     [('duck', {'defval': 'mallard', 'ptype': 'str', 'doc': 'Duck type.'})]
+                     ),
+                ),
+            }
+            name = 'foobar'
+            return ((name, modl), )
+
+        @s_module.modelrev('foobar', 201707210101)
+        def _testRev1(self):
+            '''
+            This revision adds the 'duck' property to our foo:bar nodes with its default value.
+            '''
+            try:
+                self.core.addPropDef('foo:bar:duck', form='foo:bar', defval='mallard', ptype='str', doc='Duck value!')
+            except DupPropName:
```

Comment:
Having the incremental revision updaters run on an empty cortex is essentially a waste :)  Also, could be dangerous since they're assuming they get to change the convention from the known previous state to the known next state ( which would no longer be true if they could end up getting a cortex at an arbitrary state due to rev0